test/src/tests-batchUploader.ts

```diff
@@ -89,6 +89,612 @@ describe('batch uploader', () => {
+            // clock.tick(1000);
+
+            // Batch 2
+            window.mParticle.logEvent('Test Event 1');
+            window.mParticle.logEvent('Test Event 2');
+            window.mParticle.logEvent('Test Event 3');
+
+            // Force an upload of events
+            window.mParticle.upload();
+
+            // Batch 3
+            window.mParticle.logEvent('Test Event 4');
+            window.mParticle.logEvent('Test Event 5');
+            window.mParticle.logEvent('Test Event 6');
+
+            // Force an upload of events
+            window.mParticle.upload();
+
+            // Reset clock so that the setTimeout return immediately
+            clock.restore();
+
+            setTimeout(() => {
+                const batchQueue = window.mParticle.getInstance()._APIClient
+                    .uploader.batchesQueuedForProcessing;
+
+                expect(batchQueue.length).to.equal(3);
+
+                expect(batchQueue[0].events[0].event_type).to.equal(
+                    'session_start'
+                );
+                expect(batchQueue[0].events[1].event_type).to.equal(
+                    'application_state_transition'
+                );
+
+                expect(batchQueue[1].events[0].data.event_name).to.equal(
+                    'Test Event 1'
+                );
+                expect(batchQueue[1].events[1].data.event_name).to.equal(
+                    'Test Event 2'
+                );
+                expect(batchQueue[1].events[2].data.event_name).to.equal(
+                    'Test Event 3'
+                );
+
+                expect(batchQueue[2].events[0].data.event_name).to.equal(
+                    'Test Event 4'
+                );
+                expect(batchQueue[2].events[1].data.event_name).to.equal(
+                    'Test Event 5'
+                );
+                expect(batchQueue[2].events[2].data.event_name).to.equal(
+                    'Test Event 6'
+                );
+
+                done();
+            }, 0);
+        });
+
+        it('should keep and retry batches in sequence if the transmission fails midway', done => {
+            // First request is successful, subsequent requests fail
+            window.fetchMock.post(urls.eventsV3, 200, {
+                overwriteRoutes: false,
+                repeat: 1,
+            });
+
+            window.fetchMock.post(urls.eventsV3, 429, {
+                overwriteRoutes: false,
+            });
+
+            window.mParticle.config.flags = {
+                eventsV3: '100',
+                eventBatchingIntervalMillis: 1000,
+            };
+
+            // debugger;
+
+            window.mParticle._resetForTests(MPConfig);
+            window.mParticle.init(apiKey, window.mParticle.config);
+            // Generates Batch 1 with Session Start + AST
+
+            // Adds a custom event to Batch 1
+            window.mParticle.logEvent('Test Event');
```

Comment:
I'm just thinking about this and only commenting here, but this is relevant for everywhere in this file.  We shouldn't have `Test Event`, then `Test Event 1, 2, 3, 4, 5, 6`.  We should have `Test Event 1-7`.

Reply:
I think `Test Event 0` might be more appropriate. :laugh: